adding pyupgrade & autoflake

leos opened the pull request:
Adding [pyupgrade](https://github.com/asottile/pyupgrade) and [autoflake](https://github.com/PyCQA/autoflake) to the linting in `pre-commit`.

Pyupgrade cleans up a lot of syntax, some of it legacy Python 2. Because many files use `from __future__ import annotations` it also converts the typehints to a more modern 3.10 syntax using `|` instead of `Optional` and names for types like `list` or `dict` so that `List` and `Dict` no longer need to be imported.

`autoflake` then automatically cleans up the no longer used imports.

This PR is almost entirely mechanical:
```
pre-commit run pyupgrade -a
pre-commit run autoflake -a
```

There were just a few typehint issues that this revealed that `mypy` then caught that had to be fixed manually:
```
dataprofiler/labelers/data_processing.py:1540: error: Incompatible default for argument "random_state" (default has type "None", argument has type "Union[Random, int, List[Any], Tuple[Any, ...]]")
dataprofiler/labelers/data_processing.py:1882: error: Incompatible default for argument "priority_order" (default has type "None", argument has type "Union[List[Any], ndarray[Any, Any]]")
dataprofiler/labelers/data_processing.py:1883: error: Incompatible default for argument "random_state" (default has type "None", argument has type "Union[Random, int, List[Any], Tuple[Any, ...]]")
dataprofiler/labelers/data_processing.py:2089: error: Incompatible default for argument "random_state" (default has type "None", argument has type "Union[Random, int, List[Any], Tuple[Any, ...]]")
dataprofiler/labelers/base_data_labeler.py:766: error: Incompatible default for argument "labels" (default has type "None", argument has type "Union[List[Any], Dict[Any, Any]]")
Found 5 errors in 2 files (checked 39 source files)
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/capitalone/DataProfiler?pullRequest=734) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/capitalone/DataProfiler?pullRequest=734) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/capitalone/DataProfiler?pullRequest=734) it.</sub>

taylorfturner commented:
@leos, for clarity on expectations, I may not get to review / approve this PR today. However, I will at least read through it and provide any initial comments. Best!

leos commented:
Thanks @taylorfturner ! No rush whatsover - this isn't blocking anything for me. Happy to make tweaks/updates as folks prefer.

taylorfturner commented:
> Thanks @taylorfturner ! No rush whatsover - this isn't blocking anything for me. Happy to make tweaks/updates as folks prefer.

Sound good -- just for clarity for everyone I will pop a `medium priority` on this PR in the mean time. Cheers!

ksneab7 commented on dataprofiler/labelers/classification_report_utils.py:
```diff
@@ -319,14 +319,14 @@ def classification_report(
             raise ValueError(
-                "Number of classes, {0}, does not match size of "
-                "target_names, {1}. Try specifying the labels "
+                "Number of classes, {}, does not match size of "
```
Is there a reason for using "format" as opposed to f strings like in above strings?

leos replied:
pyupgrade tries to clean up what it can - it's not 100% perfect. In this case it would make sense to turn the whole string into an f-string but it is figuring out that using `{0}` and `{1}` is redundant.

Happy to make a manual edit here to turn this into an f-string if you prefer.

JGSweets replied:
I'm okay either way for now, but I am more a fan of being consistent.

leos replied:
No problem - I've fixed these. Others can be fixed opportunistically. FYI, pylint is configured to flag strings that can be f-strings - there's actually a lot of pylint warnings/suggestions that are flagged by the current config, they all just show up by default in VS Code.

micdavis commented:
Other than Kenny's comment, LGTM. Here is a summary of changes that occur. Please call out any that I missed:

### Lowercase for certain typing such as dict and list:
`def label_mapping(self) -> Dict:` -> `def label_mapping(self) -> dict:`
`def labels(self) -> List[str]:` -> `def labels(self) -> list[str]:`

### Optional type changes:
`def preprocessor(self) -> Optional[data_processing.BaseDataPreprocessor]:` -> `def preprocessor(self) -> data_processing.BaseDataPreprocessor | None:`

### Union typing changes: 
`def set_labels(self, labels: Union[List, Dict]) -> None:` -> `def set_labels(self, labels: list | dict) -> None:`

### Remove subclass from super: 
`return super(CSVData, self)._get_data_as_records(data)` -> `return super()._get_data_as_records(data)`


### Removal of “r” from open() calls:
` with open(input_file["path"], "r") as fp:` -> `with open(input_file["path"]) as fp:`

### Removal of additional parenthesis:
```
rstr = float(
    (
        header_check_list[i].count("str")
        + header_check_list[i].count("upstr")
        + header_check_list[i].count("none")
    )
)
```
->
```
rstr = float(
    header_check_list[i].count("str")
    + header_check_list[i].count("upstr")
    + header_check_list[i].count("none")
)
```

### Formatted string change:
`expected_error = ["{}.sample_ratio must be a float.".format(optpth)]` -> `expected_error = [f"{optpth}.sample_ratio must be a float."]`

### Remove six from tests:
`six.assertCountEqual(self, ["nan"], profile.null_types)` -> `self.assertCountEqual(["nan"], profile.null_types)`

leos commented:
@micdavis there's a few more
### Get rid of old style class definition
`class SpreadSheetDataMixin(object):` -> `class SpreadSheetDataMixin:`

### Remove unnecessary future imports
`from __future__ import absolute_import, print_function` -> ` `

### Explicit raw string definitions
`"header \(0 based index\)",` -> `r"header \(0 based index\)",`

### Replace undocumented assert methods
`self.assertEquals` -> `self.assertEqual`

leos commented:
Just checking in - if folks can approve the PR if they're ok with it that would be great. Given that it touches a lot of files it risks diverging/conflicts the longer it sits unmerged.

Alternatively if folks would prefer another approach (or not adding pyupgrade) then happy to close this out.

JGSweets commented:
> Just checking in - if folks can approve the PR if they're ok with it that would be great. Given that it touches a lot of files it risks diverging/conflicts the longer it sits unmerged.
> 
> Alternatively if folks would prefer another approach (or not adding pyupgrade) then happy to close this out.

I'm going to make an intent to review Friday. Are you able to update the branch in the meantime? Thanks!

leos commented:
Thanks @JGSweets - yup, I'll update the branch before Friday

taylorfturner commented:
@leos, by way of update, I will review this tomorrow. CC @JGSweets

taylorfturner commented on dataprofiler/labelers/base_data_labeler.py:
```diff
@@ -5,7 +5,7 @@
 import sys
 import warnings
-from typing import Dict, List, Optional, Type, Union, cast
+from typing import cast
```
`flake8` caught this and `autoflake` modified. Still concerned with the other changes that `autoflake` is making

taylorfturner reviewed (commented):
Comments and with the removal of `six` in testing suite, we would also need to remove `six>=1.15.0` from `requirements.txt`

taylorfturner commented on dataprofiler/tests/profilers/test_base_column_profilers.py:
```diff
@@ -165,7 +163,7 @@ def test_combine_unqiue_sets(self):
         b = [3, 1, 4, -1]
         c = utils._combine_unique_sets(a, b)
-        six.assertCountEqual(self, [1, 2, 3, 4, -1], c)
+        self.assertCountEqual([1, 2, 3, 4, -1], c)
```
still has `import six` in the file

leos replied:
`autoflake` by default only catches unused imports that come from standardlib. Out of the box cleaning up all seemingly unused imports can be risky because they may be relied on as side effects. `autoflake` can be used with the following settings to adjust the default behavior:
```
 --imports IMPORTS     by default, only unused standard library imports are removed; specify a comma-separated list of additional modules/packages
  --expand-star-imports
                        expand wildcard star imports with undefined names; this only triggers if there is only one star import in the file; this is skipped if there are any uses of `__all__` or `del` in the file
  --remove-all-unused-imports
                        remove all unused imports (not just those from the standard library)
```

taylorfturner replied:
i see

taylorfturner commented on dataprofiler/tests/profilers/test_graph_profiler.py:
```diff
@@ -1,8 +1,5 @@
-
 import os
 import unittest
-from cgi import test
```
@JGSweets was this an important part specifically for graph somehow, though?

JGSweets replied:
@taylorfturner My assumption is this was left over from a previous test looking through visual output reports at one point for verification.

I think it is fine to remove since we aren't using it.

leos commented on .pre-commit-config.yaml:
```diff
@@ -55,3 +55,16 @@ repos:
+      - id: pyupgrade
+        args: ["--py37-plus"]
+  # Autoflake - cleanup unused variables and imports
+  - repo: https://github.com/PyCQA/autoflake
+    rev: v2.0.0
+    hooks:
+      - id: autoflake
+        args:
+          - "--in-place"
```
Sorry, yes, this is from autoflake - it is in their docs as the last line of their [Introduction](https://github.com/PyCQA/autoflake#introduction:
```
autoflake also removes useless pass statements by default.
```
This can be disabled trivially in the pre-commit config (`--ignore-pass-statements` or `--ignore-pass-after-docstring` as args), let me know if you'd like me to do that.

taylorfturner replied:
ah there it is -- I think I'd prefer to leave pass in for now. So let's add those flags in the `pre-commit` section for `autoflake`.

leos replied:
No problem - `--ignore-pass-statements` or `--ignore-pass-after-docstring` or both?

taylorfturner replied:
For now, maybe let's just do `--ignore-pass-statements`. Thx

JGSweets replied:
I'm good either way. I approved the PR for now even though this is not in yet.

I'll re-review quickly after this change has been made.

leos commented on dataprofiler/data_readers/data_utils.py:
```diff
@@ -2,7 +2,6 @@
 import json
 import re
 import urllib
-from builtins import next
```
Just to clarify, dataprofiler is long past python2 compatibility, right?

taylorfturner replied:
Yes, the minimum we focus on supporting right now is 3.8 I believe

leos replied:
Cool, if that's the case, we can bump up the syntax version in the `pre-commit` config to `--py38-plus`. Might make sense to do that as a separate PR to minimize the noise.

taylorfturner replied:
yeah agree

leos commented on dataprofiler/labelers/char_load_tf_model.py:
```diff
@@ -198,12 +196,12 @@ def load_from_disk(cls, dirpath: str) -> CharLoadTFModel:
         # load parameters
         model_param_dirpath = os.path.join(dirpath, "model_parameters.json")
-        with open(model_param_dirpath, "r") as fp:
+        with open(model_param_dirpath) as fp:
```
`pyupgrade` does it because "r" is the default mode, so while it's explicit, it's redundant.

taylorfturner replied:
yeah that's fair, I'm good with this diff

leos commented on dataprofiler/labelers/classification_report_utils.py:
```diff
@@ -319,15 +319,14 @@ def classification_report(
-                "Number of classes, {0}, does not match size of "
-                "target_names, {1}. Try specifying the labels "
-                "parameter".format(len(labels), len(target_names))
+                f"Number of classes, {len(labels)}, does not match size of "
+                f"target_names, {len(target_names)}. Try specifying the labels "
+                "parameter"
```
The proposed change will add a lot of spaces to the final string (ie. between "of" and "target_names") - is that what we want here?

taylorfturner replied:
no, not looking for a bunch of spaces ... I've never seen it render with all a bunch of spaces when formatting it like that, but I totally could be mistaken in this situation.

leos replied:
```
In [4]: a = """asdf
   ...:        cats
   ...:        dogs"""

In [5]: a
Out[5]: 'asdf\n       cats\n       dogs'
```

Multiline strings don't do anything smart with indenting unfortunately.

leos replied:
Or specifically in a function call:
```
In [11]: print("""asdf
    ...:        cats
    ...:        dogs""")
asdf
       cats
       dogs

In [12]: print("asdf "
    ...:       "cats "
    ...:       "dogs")
asdf cats dogs
```

taylorfturner replied:
I was mistaken. I'll close this comment and keep as original in your PR

leos commented:
> we would also need to remove `six>=1.15.0` from `requirements.txt`

Yup - I was planning to do that as a followup PR. Remove all imports of `six` and then remove it from the requirements, running `pipdeptree` to remove any transitive deps it left behind as well.

JGSweets commented on .pre-commit-config.yaml:
```diff
@@ -55,3 +55,16 @@ repos:
         'python-dateutil', 'pytz', 'six', 'pyarrow', 'chardet', 'fastavro',
         'python-snappy', 'charset-normalizer', 'psutil', 'scipy', 'requests',
         'networkx','typing-extensions']
+  # Pyupgrade - standardize and modernize Python syntax for newer versions of the language
+  - repo: https://github.com/asottile/pyupgrade
+    rev: v3.3.0
+    hooks:
+      - id: pyupgrade
+        args: ["--py37-plus"]
```
Should this be `py38-plus`? We can always update in a follow-up PR so we don't have to re-review everything again.

leos replied:
Yup, will do that in a followup PR

leos replied:
Actually, just did it here, it only had one tiny change - https://github.com/capitalone/DataProfiler/pull/734/commits/d6afd70347406a028effabf82438a1b3b7ee392b - moving the import for `Protocol` from `typing_extensions` to `typing.

JGSweets commented:
@leos  Thank you for working on this and giving us time to review it in full. Definitely a worthwhile bit of cleanup!

JGSweets commented on dataprofiler/data_readers/avro_data.py:
```diff
@@ -58,7 +58,6 @@ def file_encoding(self, value: Any) -> None:
 
         Required by mypy because the inherited self.file_encoding is read-write.
         """
-        pass
```
To be fair, this code here could use a clean-up as it is not ideal to have a function such that we need to pass mypy.

leos commented on dataprofiler/labelers/regex_model.py:
```diff
@@ -250,11 +244,11 @@ def predict(
             if verbose:
                 sys.stdout.flush()
-                sys.stdout.write("\rData Samples Processed: {:d}   ".format(i + 1))
+                sys.stdout.write(f"\rData Samples Processed: {i + 1:d}   ")
```
This would be a functional change - the formatting syntax (ie. `:d` is the same in f-strings as it is with `format()`.

If y'all would like me to do that, happy to.

JGSweets replied:
Correct, `:d` is already formatting.

Let's leave it as is

leos commented:
Addressed all PR feedback:
* Restored the unnecessary `pass` statements and disabled `autoflake` from removing them
* Moved `pyupgrade` to `py38-plus`

https://github.com/capitalone/DataProfiler/pull/734/files/e7e259443304917f0c8d6debe3a10abac5a35468..d6afd70347406a028effabf82438a1b3b7ee392b

Let me know if there's anything else before this can land.

taylorfturner reviewed (approved):
LGTM! Thanks for the contribution and bearing with us on the timeline. Cheers!
